llvm/lib/Transforms/IPO/MergeFunctions.cpp

```diff
@@ -891,8 +891,10 @@ bool MergeFunctions::writeThunkOrAlias(Function *F, Function *G) {
 
 // Merge two equivalent functions. Upon completion, Function G is deleted.
 void MergeFunctions::mergeTwoFunctions(Function *F, Function *G) {
-  if (F->isInterposable()) {
-    assert(G->isInterposable());
+  if (F->isInterposable() || G->hasWeakODRLinkage() ||
```

Comment:
I'm not familiar with this code - what is this guarding? And is it intended that you are checking F in the first check but G in the newly added checks?

Do we need to worry about AvailableExternallyLinkage? ThinLTO converts non-prevailing linkonce_odr and those won't stick around as out of line copies, so might be ok as I think the infinite loop you are describing can't happen in that case?

Reply:
If I read correctly, this is going to happen with or without LTO, and having the prevailing symbol info doesn't help here. Creating the infinite loop is one thing, but having one call the other (if either of them is interposable) will break the runtime since providing a runtime replacement will cause either no replacement or replacing one function will cause replace both functions. In this case, you can only use the thunk to merge them.

So I think we just just check here:
```
if (F->isInterposable() || G->isInterposable())
```

Reply:
I see, it is rely on ODR to assume if the replacement happens, it is functionally equivalent so it is safe.

Maybe it is safer to replace all references of `isInterposable()` in this pass with the stricter definition here that includes ODR.

Reply:
Unfortunately replacing all reference of `isInterposable` gives worse results, in particular, if we have a function `F` that is not interposable and not linkonce/weak ODR and another function `G` that is link once/weak ODR, then `G` should be able to call `F`.

I also updated the comparator to order ODR functions before non-ODR functions to consistently support the case above.